rfc/006-genpolicy-platforms.md

```diff
@@ -0,0 +1,111 @@
+
+### The agent
+
+The functionality in the agent is very simple: an in-memory policy engine is equipped with the incoming policy, and subsequent agent requests are subject to policy validation.
+When a container is started with the tardev-snapshotter, the agent also ensures that the layers are mounted with `dm-verity` protection.
+
+## State of genpolicy
+
+This is a snapshot of the features in Microsoft's fork that are yet to be implemented in upstream Kata: <https://gist.github.com/burgerdev/9330236c209f49c9d97e01f278b3a915>.
+Setting aside the changes for Azure's CSI drivers, the two main categories are YAML support and hardening.
+Both of these are desirable to backport, the difficulty stems from code layout changes between the repositories.
+
+Contrast uses the Microsoft fork of `genpolicy` and its configuration, with a few patches on top.
+
+The upstream policy has support for `dm-verity` calculation, but the corresponding rule is commented out.
+It also uses an outdated and unmaintained version of `tarindex`.
+Upstream will need to introduce support for other snapshotters (that is, Nydus) eventually.
+
+## Proposal
+
+Contrast shouldn't need to package different versions of the `genpolicy` tool or the CLI.
+Instead, the upstream tool should be able to cover generation for all target platforms.
+Contrast should bundle rules and settings for its target platforms, and invoke the tool accordingly.
+
+How do we get there?
+
+### Focus on upstream
+
+The Kata `genpolicy` is already close to the desired state, and will get even more traction once the CoCo community needs policy enforcement.
+We can accelerate progress by porting necessary changes ourselves.
+Also, contributions from Contrast can ensure that the tool becomes flexible enough to support our use cases.
+
+### Prepare to maintain patches
+
+We would like to support a mix-and-match of snapshotters and target platforms, but the CoCo community consensus is to focus on one.
+Thus, it's likely that Contrast will need patches, for example for an updated `tarindex` or even the entire `dm-verity` calculation.
+
+### Develop our own rules and settings
+
+Due to the nature of the settings file, Contrast can't rely on a single vendor to provide this.
+On the other hand, we don't want to force our users to make modifications for our standard target platforms.
+Thus, we develop settings for our known target platforms, validate them through CI and bundle them with the CLI.
+
+The upstream rules should eventually be compatible with all non-tardev platforms.
+This is a shared interest throughout the wider CoCo community, but we can accelerate it through contributions.
+
+For the tardev snapshotter platforms (that is, AKS for now), we start with the Microsoft rule set.
+Once the necessary patches grow uncomfortable (for example reference value divergence), we port the missing parts to a patch for the upstream rules.
```

Comment:
Just for my understanding:
In other words, this means bringing all the msft changes to upstream except the tardev support.
Leaving tardev as the only difference/reason for the existence of the msft fork?

Reply:
Microsoft has other patches that might not make it upstream, e.g. for their CSI drivers. My idea here is that we would port these so that they _would_ apply to upstream Kata, but apply them in the Contrast build. However,  this may not become necessary if upstream is "compatible enough".

Reply:
Upstream would not accept these patches since they're specific for Azure's CSI-Plugins in this case?
What happens if we support other platforms such as GCP or AWS.
Would we have patches for things such as CSI for all of them?

Reply:
I'm not sure what Kata's strategy is wrt. insecure volume mounts in confidential containers, but I don't expect it to be a priority and it may thus take considerable time until adoption.

Neither do I think we should make it _our_ priority, because there's just not a good use case that can't be addressed with block storage, which is much easier to check with policies.